c10/cuda/CUDADeviceAssertion.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include <c10/cuda/CUDAException.h>
+#include <c10/macros/Macros.h>
+
+// Used by preprocessor to convert things to strings
+#define C10_CUDA_DEVICE_ASSERTION_STRINGIFY(x) #x
+#define C10_CUDA_DEVICE_ASSERTION_TOSTRING(x) \
+  C10_CUDA_DEVICE_ASSERTION_STRINGIFY(x)
+#define C10_CUDA_DEVICE_ASSERTION_LINE_STRING \
+  C10_CUDA_DEVICE_ASSERTION_TOSTRING(__LINE__)
+
+namespace c10 {
+namespace cuda {
+
+#ifdef TORCH_USE_CUDA_DSA
```

Comment:
Err, do we have to feature gate this? Would be better if we can just turn this on unconditionally everywhere.

Reply:
@ezyang  Having the feature gate is useful for disabling this feature on older Nvidia GPUs and most AMD GPUs. The older Nvidia GPUs might not be an issue, but the AMD GPUs might. We can try without the gating, if you prefer.